### ISSUE TITLE
Magma Bazel Show&Tell + Q&A (2023-02-15)

### ISSUE THREAD

**Issue by LKreutzer** [magma/magma#15019]
# Magma AGW with Bazel

Magma Bazel Show & Tell - 15th Feb. 2023

The Youtube recording of this session will be available on the [meeting page](https://wiki.magmacore.org/pages/viewpage.action?pageId=12257079).

 [Link to the Youtube recording](https://www.youtube.com/watch?v=1-TzdbMYfIU).

## What is Bazel and what are the benefits for Magma?

Some of the benefits:
  - multi-language, multi platform, extensible system
    - ideal for multilingual mono-repos like Magma
    - can support multiple architectures, configurations etc.
    - possible to extend to other Magma components in the future
      - ❗ this is a non-goal for now
  - free open-source project, with active community
    - setup not Magma specific -> can also find outside help
  - clear file and target structure
    - already community contributions at Magma that modify Bazel, e.g. [#14966](https://github.com/magma/magma/pull/14966), [#14494](https://github.com/magma/magma/pull/14494), [#14489](https://github.com/magma/magma/pull/14489), [#14449](https://github.com/magma/magma/pull/14449), [#14436](https://github.com/magma/magma/pull/14436), [#14420](https://github.com/magma/magma/pull/14420), [#14356](https://github.com/magma/magma/pull/14356), [#14339](https://github.com/magma/magma/pull/14339), [#14290](https://github.com/magma/magma/pull/14290), ...
  - reproducible builds
  - hermetic (sand-boxed) execution with strictly defined in- and outputs
  - build [profiling](https://bazel.build/rules/performance#performance-profiling) and metrics
    - easy to get information, e.g. about performance bottle necks for further optimization 
  - powerful [query language](https://magma.github.io/magma/docs/next/bazel/agw_with_bazel#querying)
    - easy to analyze service architecture, create dependency graphs etc.
  - granular targets and dependencies, no circular dependencies allowed, strict visibility control
    - helps to avoid regression and keep services separated
    - know which unit tests need to run for any change
    - only build and test what is really needed
  - incremental builds
    - faster iteration and development
  - [(remote) caching](https://bazel.build/remote/caching)
    - faster builds and CI
  - parallelization and scalability, remote execution, [distributed builds](https://bazel.build/basics/distributed-builds)

## Where can I find further information?

  - Magma Bazel docs: [AGW with Bazel](https://magma.github.io/magma/docs/next/bazel/agw_with_bazel)
  - Official Bazel docs: [bazel.build](https://bazel.build/reference)
  - Magma Slack: [#bazel](https://magmacore.slack.com/?redir=%2Farchives%2FC033S1CEAUU)
  - Official Bazel slack: [slack.bazel.build](https://slack.bazel.build/)

## How does [Bazel caching](https://magma.github.io/magma/docs/next/bazel/agw_with_bazel#caching) work?

  - Remote caching
    - Recommended method, scalable, convenient, robust 
    - Simplest setup is with [Google Cloud](https://bazel.build/remote/caching#cloud-storage)
      - `--remote_cache=https://storage.googleapis.com/bucket-name`
      - `--google_credentials=/path/to/your/secret-key.json` 
    - See [Magma docs](https://magma.github.io/magma/docs/next/bazel/agw_with_bazel#bazel-remote-cache) on [bazel-remote](https://github.com/buchgr/bazel-remote/) setup with AWS/Terraform.
  - [Docker cache](https://magma.github.io/magma/docs/next/bazel/agw_with_bazel#docker-bazel-cache)
    - Pre-built cache in Docker images
    - Published weekly, configured in `.github/workflows/docker-builder-devcontainer.yml`.
    - Three configs
      - [bazel-cache-plain](https://github.com/magma/magma/pkgs/container/magma%2Fbazel-cache-plain): `ghcr.io/magma/magma/bazel-cache-plain:latest`
      - [bazel-cache-prod](https://github.com/magma/magma/pkgs/container/magma%2Fbazel-cache-prod): `ghcr.io/magma/magma/bazel-cache-prod:latest`
      - [bazel-cache-asan](https://github.com/magma/magma/pkgs/container/magma%2Fbazel-cache-asan): `ghcr.io/magma/magma/bazel-cache-asan:latest`
   - GitHub cache (see [#14988](https://github.com/magma/magma/pull/14988)), etc.

## CI runtime comparison

⚠️ The following numbers are rough estimates taken from ~30 workflow runs:

|   | Make  | Bazel  |
|---|---|---|
| Bazel workflow runtime on PRs  |   |  1-3 min (no C/C++ changes) |
| Bazel workflow runtime on PRs  |   |  5-26 min (with C/C++ changes)  |
| ...  |   |   |
| All AGW unit tests  |   | 8-10 min (outdated cache 20 - 35 min)  |
| MME unit tests  | 38-48 min  |   |
| LTE python tests  | 6-7 min  |   |
| Session manager tests  | 9-12 min  |   |
| ...  |   |   |
| C/C++ coverage  | 50-58 min | 13-26 min |
| ...  |   |   |
| Magma Debian package build  | 80-104 min  | 7-14 min |

## Magma CI with Bazel

- [GitHub Actions Workflows](https://magma.github.io/magma/docs/next/bazel/agw_with_bazel#workflows)
  - How to see which test failed and why? 
- [Bazel-diff](https://magma.github.io/magma/docs/next/bazel/agw_with_bazel#bazel-diff)
  - How to know which tests need to run in CI?
  - Do the Debian packages need to be built?
  - Have Python services been affected?

## What Bazel files are there and what are targets?

- [Starlark configuration language](https://github.com/bazelbuild/starlark)
  - Deterministic, hermetic, parallelized Python dialect
- [Bazel file structure of the Magma AGW](https://magma.github.io/magma/docs/next/bazel/agw_with_bazel#structure-and-files)
- [Bazel targets](https://magma.github.io/magma/docs/next/bazel/agw_with_bazel#targets)

## What [Bazel commands](https://magma.github.io/magma/docs/next/bazel/agw_with_bazel#commands) are there?

- [`bazel help`](https://bazel.build/docs/user-manual#help)
- How to [build](https://magma.github.io/magma/docs/next/bazel/agw_with_bazel#building) a library/service/... ?
  - Where are the artifacts?
    - `magma/bazel-bin/`
    - The exact path is also contained in the Bazel logging. 
- How to [build the Magma Debian package](https://magma.github.io/magma/docs/next/bazel/agw_with_bazel#release-build)?
  - The relevant targets are defined in `lte/gateway/release/BUILD.bazel`. 
- How to run [tests with Bazel](https://magma.github.io/magma/docs/next/bazel/agw_with_bazel#testing)?
  - [Unit tests](https://magma.github.io/magma/docs/next/bazel/agw_with_bazel#unit-tests)
  - [Integ tests](https://magma.github.io/magma/docs/next/bazel/agw_with_bazel#lte-integration-tests)
  - [Python sudo tests](https://magma.github.io/magma/docs/next/bazel/agw_with_bazel#python-sudo-tests)
  - [Load tests](https://magma.github.io/magma/docs/next/bazel/agw_with_bazel#load-tests)

## How to add a new unit test (or a library/module/service/...)?

- How to write a [`BUILD.bazel` file](https://bazel.build/concepts/build-files)
  - See also the documentation e.g. on the [Python rules](https://bazel.build/reference/be/python).
- [How to add dependencies](https://magma.github.io/magma/docs/next/bazel/agw_with_bazel#dependencies)
- Live coding demo: Add a smsd unit test and bazelify it.

## What is the current state of the Make-to-Bazel migration of the AGW?

- Everything that was possible with Make is now also possible with Bazel.
- A couple of CI testing workflows are already completely migrated and only use Bazel. This includes all AGW unit tests and the coverage generation.
- The Magma Debian package build is currently duplicated, i.e. Make created artifacts are created (tagged as 1.8.0) and Bazel artifacts (tagged as 1.9.0).
- The respective integration tests on Debian artifacts are also duplicated.

### What is still in progress?

- The FEG integration tests are still running on an environment that is based on an AGW build with Make.
  - A pull request ([#14852](https://github.com/magma/magma/pull/14852)) to change this to an AGW installed via a Debian package is in review.

### Which tasks need to be taken over by the community?

- [List of open Bazel issues](https://github.com/magma/magma/issues?q=is%3Aopen+is%3Aissue+label%3Abazel)
  - Bazel tasks are tagged with the label `bazel`.
  - Most of these issues are not necessary for a full AGW switch-over.
- Changes that can be done only once a switch-over has happened, e.g.
  - https://github.com/magma/magma/issues/12755
  - https://github.com/magma/magma/issues/11959
  - https://github.com/magma/magma/issues/13807
- Usual build system maintenance and optimization
  - E.g. upgrade Bazel version, see [github.com/bazelbuild/bazel/releases](https://github.com/bazelbuild/bazel/releases). 
    - See also https://github.com/magma/magma/issues/14791.

**Comment by LKreutzer** [comment]
This issue is only to document the meeting notes for the show and tell session.
